### DISCUSSION TITLE
Add support for bluetooth/local wifi messaging

### DISCUSSION THREAD

**Discussion by nobaraos12** [simplex-chat/simplex-chat#1501] [Ideas]
Feature request for adding support for offline end to end encrypted bluetooth messaging or even local wifi

Use cases 

**- airline travel (airlines only allow imessage and whatapp on free wifi as they can monitor)**
- backcountry travel (no cellular service)
- conjested networks (helps relatively close communication when network speed is slow or unreliable) 

Update:
According to my recent search there is another app that does bluetooth messaging called briar https://briarproject.org/
this may help development. Note they seem to only have support for android and linux

**Comment by epoberezkin** [comment]
look forward to seeing how much interest there is - vote above!

**Reply by nobaraos12** [reply]
I think the people have spoken

**Reply by Darin755** [reply]
> look forward to seeing how much interest there is - vote above!

I feel like this very much a case of easier said than done. With that being said, have you looked into gossip protocols? I feel like that would be the only way to implement this. It would likely need to be Bluetooth since Wifi direct would need user interaction.(at least on Android)

The way this could work is that nodes could randomly share what they have with other nodes with the contents being completely encrypted. In the app it could just decrypt the stuff it knows how to. You would need some sort of Anti Entropy protocol. https://www.geeksforgeeks.org/anti-entropy-in-distributed-systems/ The advantage of sending and receiving everyone's massages is that it makes it harder to track individual recipients.
 
 The reason I am giving this suggestion is that I am working on a totally separate project that uses Gossip protocols.

**Comment by Snake883** [comment]
Some other use cases:

- Natural Disaster
- War
- Protests
- Offgrid
- Search and Rescue
- Poor reception areas/buildings
- Communicate with neighbors when infrastructure isn't available.

Would be even better if it relayed or meshed.
Even better if there was a microblog/twitter type feature.
P2P has appeal.
IPFS for distribution.

**Reply by nobaraos12** [reply]
another great use case is in heavily dense crowds trying to find another individual when service is unavailable.
Example is ski resort lines, no service and dense crowds

**Reply by nobaraos12** [reply]
> Would be even better if it relayed or meshed. Even better if there was a microblog/twitter type feature. P2P has appeal. IPFS for distribution.

I completely agree, similar to how telegram works

**Reply by d-RLY** [reply]
I really really want to see more messaging apps have mesh support for all the above reasons.  One major feature I would be super happy to see would be the mix of bouncing messages around all possible routes.  And know that even if I have zero cell or internet that my message will re-route to the next person close to me and jump to network if they do have service or is on wifi I can't get onto.  Though for the purpose of not eating up large amounts of other people's data if they are the last hop before getting onto the internet or cell data.  It may be helpful to only send text in that situation and have an option (or a different "send" button) to specify when a "bounce" message is being sent.  Or have the app ask if you would like to strip media from the message if the person you are trying to message isn't within bluetooth or local network wifi and your normal service isn't giving a signal (or a signal below a certain amount or if you could toggle some kind of "emergency mode").  P2P could really be used for good and safety.  Of course it could be even better if data prices weren't a thing since stuff could much more easily be distributed and there wouldn't be an issue of fucking up someone else's data caps.

**Comment by serrq** [comment]
I am very happy to share this topic with you, because I like it.

Personally I don’t have much enthusiasm and trust in bluetooth as it is controlled by a private consortium and it is vulnerable to many malicious attacks. But I could be wrong. So I leave this consideration to those who deal with it for a living.

The use case that I present is speleology, and I would like to use SimpleX with the wifi chip integrated (and soon through the 802.11ah or HaLow standard) without the use of a router coordinator in a way that would resemble that used in the military and civil systems gotenna.

https://www.wi-fi.org/discover-wi-fi/wi-fi-certified-halow

https://www.morsemicro.com/

https://gotenna.com/

https://gotennamesh.com/products/mesh

https://gotennapro.com/

I think they exploit a mesh network, totally offline (no internet).

What I propose is to turn our devices at need into nodes, so that we can reach a fellow (a speleologist or a mountaineer) of roped rope that maybe fell into a cliff 10 meters below us and need for help (communication for example).

**Reply by mika-nl** [reply]
I can see this for myself as well.
I've had the idea before that an airdrop feature would be nice this way.

Imagine you are in a meeting, party, concert, etc.
You have taken pictures or a document that you want to share with the people (not contacts) in the same room.
The "airdrop" would then be a group where you can share text and files offline via wifi or bluetooth . but the same group also via internet so you can keep in touch later with the people in the group that you have met at the meeting / party / concert / and so on.

**Comment by trymeouteh** [comment]
I like how this app uses relay nodes and allow you to connect to multiple nodes at once which is quite censorship resistant.

Adding P2P connectivity would also be great in the event the internet is down. If the internet infrastructure goes down or goes offline which is done in some countries were they shutdown the internet, P2P should be a fallback.

P2P will require the LAN networks to open up ports for SimpleX to work P2P. P2P in Simplex P2P setting should include the option to use UPnP, choosing what port number to use, etc. And to have a setting to use the same P2P setting for all profiles or use different P2P settings for each profile for additional privacy.

Again, P2P will have to be a fallback and will likely need to be configured by the user to some extent. Otherwise SimpleX should use relay servers for reliable messaging.

**Reply by Darin755** [reply]
Simplex could always use i2p or tor

**Comment by LinuxinaBit** [comment]
I would love to see the P2P and offline features of Briar combined with the amazing privacy features of SimpleX.
Berty's IPFS-based protocol could be a good starting point, but it would need to integrate with SimpleX's stuff.

Bluetooth or BLE mesh would be awesome, maybe even Ad-hoc WiFi networks alongside that on supported devices.
It could be able to work with other protocols too, like LoRa with the PineDio or something similar, but it shouldn't *require* you to buy separate devices or antennas which may not be available.
Of course because of various security concerns and such, it should be toggle-able, or maybe even opt-in.

**Reply by Darin755** [reply]
I don't really trust ipfs when it comes to privacy and security. Its a fairly new technology that needs validation before it is put into production

**Reply by LinuxinaBit** [reply]
I misread Berty's documentation, they actually use something similar to Briar for offline communication.
It would probably be better to start off with Briar's implementation because of that anyway...

**Comment by Wjxfi** [comment]
This is a very cool idea! Look at how the Meshenger app did it below. It did it perfectly. 

https://www.f-droid.org/packages/d.d.meshenger/

**Reply by LinuxinaBit** [reply]
Keep in mind that Meshenger is just over a LAN connection, BLE mesh alongside that would be ideal.

**Reply by Darin755** [reply]
The problem with Bluetooth is that it is much slower than WiFi or internet. You could do text but audio is out of the question unless two devices are close together. (Think Bluetooth speaker)

**Reply by LinuxinaBit** [reply]
Why do you think Briar compresses the heck out of its images?
Something that would wait to send audio/full resolution images until internet/LAN connected, but still send text and possibly heavily compressed image previews over Bluetooth, would definitely be necessary.

**Comment by Yiannis128** [comment]
It would be helpful if a relay app was also available on Android and IOS that allowed relays to be set up locally through local wifi and Bluetooth

**Comment by Wjxfi** [comment]
Any news ?

**Comment by Darin755** [comment]
This could be implemented though wifi mesh. Android does not support 802.11s but it does support Ad-Hoc

Maybe something similar to this: https://github.com/UstadMobile/Meshrabiya

**Comment by Samillion** [comment]
Hear, hear. LAN/Wifi support would be wonderful. Especially useful in areas with unstable internet but the local network works fine.

**Comment by User1l0** [comment]
While briar messanger app have this feature a long time ago and it works perfectly. First I'll recommend the devs of simplex to use or inspect the briar project's messenger source code to learn briar's offline bluetooth/wifi mode features. To easly add this feature to Simplex.

**Reply by Darin755** [reply]
That is probably not a good idea as briar is under a separate license. Briar isn't all that unique from a technical perspective but it is pretty powerful as a concept. What intrigues me about Briar is how they utilize contacts to create levels of trust. On Briar you can share a blog and it will be synced with other Briar devices. For private group chats messages are only transmitted to users in the group. 

From a Simplex Chat perspective this could just be a group chat where messages are sent to recognized peers. This would be good from a trust perspective because you would only be routing messages that are also intended for yourself.

![briar graphic](https://briarproject.org/img/howitworks3.svg)

**Comment by Darin755** [comment]
Would Simplex Chat also support Wifi direct and local lan connections? I know under Android Wifi direct should provide a lot more range and speed. It wouldn't work for Apple devices but I don't think there is an easy way to get around that. Apple devices do support mDNS so it should be possible to connect to existing networks for local messaging.

**Comment by droptile** [comment]
This use case is very niche. As mentioned earlier there are other FOSS solutions to local messaging like Briar and Meshtastic. It would be better for SimpleX to be the best secure/private chat app in the world than be a mess of multiple different chat protocols. This is the kind of rabbit hole that Element/Matrix fell into. I can think of a dozen different features that would have more appeal and get more users to use SimpleX than this.

Interesting concept and the use case is valid but I don't think it warrants the time and cost of implementation compared to other features at this time.

**Reply by Darin755** [reply]
Honestly what we really need is a standard protocol or network for doing local p2p connections. Such a protocol could be used for all sorts of things such as p2p messaging and find my device.

**Reply by LinuxinaBit** [reply]
@Darin755 if you want it to scale, a dedicated protocol like Reticulum is what you want… It’s hard to adapt an existing app like SimpleX to do something like that.

**Comment by LP97469** [comment]
It would be nice if slow message passing over multiple nodes would be possible, as in:
1. I create a message meant for Alice.
2. I pass the message to those around me they keep it for a while
3. someone runs into Alice at some point in the future and she with our shared secret decrypts the message I originally sent
